### ISSUE TITLE
Guidelines on implementing classes for multi-threading based parallel compilation

### ISSUE THREAD

**Issue by czgdp1807** [lcompilers/lpython#1151]
If we want to follow multithreading based parallel compilation then we should follow some simple guidelines,

1. **Keep everything (command line options, list API objects, dict API objects, passes) local to a visitor and launch separate visitor instances in separate threads.** This way we avoid locking overall. Since we create the visitor instance inside a function call so each thread will call that function and hence the scope of each thread will be isolated from the other.
2. **Don't use static variables in any class. Always keep all the variables local to an object because each thread will have its own objects isolated from others.** Currently `SymbolTable` and `ASRUtils::LabelGenerator` classes use static variables
3. **Avoid singleton classes unless extremely necessary.** They have to be safeguarded by locks which hinders performance. Currently, `ASRUtils::LabelGenerator` is a singleton class and we should make it local object. I will do it anyways when other priorities are complete.

We should include these guidelines in the README.md or the CONTRIBUTING.md file

**Comment by certik** [comment]
I think so. There might be more things. In general, we need to keep the code as "state free" as possible. If we have to maintain state, let's do it inside a class. Make sure two visitors can run at the same time, and instantiated at the same time. Etc.

**Comment by czgdp1807** [comment]
> There might be more things. In general, we need to keep the code as "state free" as possible. If we have to maintain state, let's do it inside a class.

Makes sense. We call functions in parallel threads so all objects should be created inside function scopes. That will solve the problem.

> Make sure two visitors can run at the same time, and instantiated at the same time. Etc.

Exactly. Do we plan to test this? Not a high priority though I think.

**Comment by certik** [comment]
I don't know how to test this in C++. We could test it once we have a C++ frontend, but that is much later.
